drivers/espi/espi_taf_npcx.c

```diff
@@ -0,0 +1,470 @@
+			cycle_type = CYC_SCS_CMP_WITH_DATA_LAST;
+			len = total_len;
+		} else {
+			cycle_type = CYC_SCS_CMP_WITH_DATA_MIDDLE;
+		}
+	} while (total_len);
+
+	return 0;
+}
+
+int espi_taf_npcx_flash_write(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr = (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)(taf_data_ptr->data);
+	struct npcx_taf_head taf_head;
+	uint16_t len;
+	int rc;
+
+	if (espi_taf_check_write_protect(dev, pckt->flash_addr,
+					 pckt->len, taf_data_ptr->tag)) {
+		LOG_ERR("Access protection region");
+		return -EINVAL;
+	}
+
+	rc = flash_write(spi_dev, pckt->flash_addr, data_ptr, pckt->len);
+	if (rc) {
+		LOG_ERR("flash write fail 0x%x", rc);
+		return -EIO;
+	}
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
+
+	taf_head.pkt_len = len + 3;
+	taf_head.type = CYC_SCS_CMP_WITHOUT_DATA;
+	taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+	taf_head.llen = len & 0xFF;
+	memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+	taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+	return 0;
+}
+
+int espi_taf_npcx_flash_erase(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr = (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)taf_data_ptr->data;
+	uint32_t addr = pckt->flash_addr;
+	uint32_t len = pckt->len;
+	struct npcx_taf_head taf_head;
+	int rc;
+
+	if (espi_taf_check_write_protect(dev, addr, len, taf_data_ptr->tag)) {
+		LOG_ERR("Access protection region");
+		return -EINVAL;
+	}
+
+	rc = flash_erase(spi_dev, addr, len);
+	if (rc) {
+		LOG_ERR("flash erase fail");
+		return -EIO;
+	}
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
+
+	taf_head.pkt_len = len + 3;
+	taf_head.type = CYC_SCS_CMP_WITHOUT_DATA;
+	taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+	taf_head.llen = len & 0xFF;
+	memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+	taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+	return 0;
+}
+
+int espi_taf_npcx_flash_unsuccess(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr
+			= (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)taf_data_ptr->data;
+	struct npcx_taf_head taf_head;
+	uint16_t len;
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
+
+	taf_head.pkt_len = len + 3;
+	taf_head.type = CYC_UNSCS_CMP_WITHOUT_DATA_ONLY;
+	taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+	taf_head.llen = len & 0xFF;
+	memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+	taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+	return 0;
+}
+
+static int espi_taf_npcx_manage_callback(const struct device *dev, struct espi_callback *callback,
+					 bool set)
+{
+	return 0;
+}
```

Comment:
is this intentional? provide the callback but uncoditionally returning success? You could also not define it at all...

Reply:
Done.